### DISCUSSION TITLE
I2C master bus goes to timeout when I2C slave is receiving data

### DISCUSSION THREAD

**Discussion by hitech95** [stm32duino/Arduino_Core_STM32#1938] [Q&A]
Hi,
I'm working on a simple code to allow a MCU to act as a gateway.
I access to the information via I2C, I'm correctly receiveing commands/registers to read and write sucessfully.

But I've noticed that if I want to use another I2C bus while I'm inside `onHandleReceive` the I2C bus fail.

```cpp
void onHandleReceive(int bytes) {
uint8_t register = bus->read();
uint8_t value = bus->read();

// More logic here to process the register and value

bus2->beginTransaction(addr);
bus2->write(buffer, len);
bus2->endTransmission(true);
}
```

Is is possible to run the two busses together?
I'm gonna use gdb to debug the calltrace to see if I can spot something.

Another solution would be to process the data in a second time.
Outside the callback.

**Comment by hitech95** [comment]
Small update:
I'm getting an `I2C_TIMEOUT` from the `i2c_master_write` function in `twi.cpp`

The I2C is ready before, but when it wait for the transfer to be completed it fails with timeout error.

**Reply by fpistm** [reply]
I guess calling I2C api from ISR is not a good idea. As I2C use IT then when you call I2C on an other instance while I2C IT is already fired then next called are simply blocked.

**Reply by hitech95** [reply]
Yes, you are right.
I've forgot that that method is called by an ISR. In fact after some brainstorming it was clear to me why I was getting the timeout...
IRQs are disabled in the ISR.

No i'm using a flag (task in reality in the scheduler i'm using) to post process the incoming data right after the ISR...
But now I think that there is no way to keep the bus busy to singnal completion back to the master.

In any case I might have an example of my solution to share later, but it uses the `TaskScheduler `library to manage the post process.
